src/registry.ts

```diff
@@ -0,0 +1,124 @@
+import { each, has, reduce, valuesIn } from 'lodash';
+import { Collector, CounterValue, HistogramValue, Metric } from './collector';
+import { Counter } from './counter';
+import { Gauge } from './gauge';
+import { Histogram } from './histogram';
+
+import { formatCounterOrGauge, formatHistogramOrSummary } from './utils';
+
+export type CollectorType = 'counter' | 'gauge' | 'histogram';
+
+type CollectorForType<T extends CollectorType> =
+  T extends 'histogram' ? Histogram :
+    T extends 'gauge' ? Gauge :
+      T extends 'counter' ? Counter :
+        never;
+
+type RegistryItem<T extends CollectorType> = {
+  [key: string]: {
+    type: T,
+    help: string,
+    instance: CollectorForType<T>,
+  },
+};
+
+export class Registry {
+  private data: {
+    [K in CollectorType]: RegistryItem<K>
+  };
+
+  constructor() {
+    this.data = {
+      counter: {},
+      gauge: {},
+      histogram: {},
+    };
+  }
+
+  private checkExists(type: CollectorType, name: string): void {
+    // check for name collisions
+    if (has(this.data, `${type}.${name}`)) {
+      throw new Error(`A metric with the name '${name}' already exists for type '${type}'`);
+    }
+  }
+
+  create(type: 'counter', name: string, help?: string): Counter;
+  create(type: 'gauge', name: string, help?: string): Gauge;
+  create(type: 'histogram', name: string, help?: string, histogramBuckets?: number[]): Histogram;
+  create(
+    type: CollectorType,
+    name: string,
+    help: string = '',
+    histogramBuckets: number[] = [],
+  ): Collector<any> {
+    this.checkExists(type, name);
+
+    let instance;
+    if (type === 'counter') {
+      instance = new Counter();
+      this.data.counter[name] = { help, instance, type };
+    } else if (type === 'gauge') {
+      instance = new Gauge();
+      this.data.gauge[name] = { help, instance, type };
+    } else {
+      instance = new Histogram(histogramBuckets);
+      this.data.histogram[name] = { help, instance, type };
+    }
+
+    return instance;
+  }
+
+  metrics() {
+    // Returns a string in the prometheus' desired format
```

Comment:
suggest using JSDoc annotation here: https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript

that way most code editors (vim, webstorm, vscode, atom, etc.) that support JSDoc will show this comment.